cleanup: move scheduler unit tests to use PodWrapper

yibozhuang opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Move to use testing `PodWrapper` where applicable to
reduce duplicating pod creation code and shorten
number of lines.

Adding additional wrapper functions in PodWrapper
to ensure it covers different pod spec under different plugins
and test cases.

Signed-off-by: Yibo Zhuang <yibzhuang@gmail.com>

#### Special notes for your reviewer:

This is a follow PR and builds on top of #109536 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

yibozhuang commented:
/assign @Huang-Wei

yuanchen8911 reviewed (approved):
Thanks for the PR. /lgtm

Huang-Wei commented:
/ok-to-test

Huang-Wei commented:
Thanks for the PR. Please fix the CI failures and I will review.

yibozhuang commented:
/retest

kerthcet commented on pkg/scheduler/testing/wrappers.go:
```diff
@@ -266,6 +266,18 @@ func (p *PodWrapper) NominatedNodeName(n string) *PodWrapper {
+
+// Conditions sets `conditions` as .status.Conditions of the inner pod.
+func (p *PodWrapper) Conditions(conditions []v1.PodCondition) *PodWrapper {
+	p.Status.Conditions = conditions
```
`p.Status.Conditions = append(p.Status.Conditions, conditions...)`

yibozhuang replied:
Thanks, fixed.

kerthcet commented on pkg/scheduler/testing/wrappers.go:
```diff
@@ -285,6 +297,17 @@ func (p *PodWrapper) HostPort(port int32) *PodWrapper {
+		p.Spec.Containers = append(p.Spec.Containers, v1.Container{})
+	}
+
+	p.Spec.Containers[0].Ports = ports
```
Why not the all Containers?

yibozhuang replied:
Changed to allow a container to be created and then assign the ports to the new container, thanks for catching this.

kerthcet commented on pkg/scheduler/testing/wrappers.go:
```diff
@@ -417,6 +446,20 @@ func (p *PodWrapper) Label(k, v string) *PodWrapper {
+}
+
+// CreationTime sets the CreationTimestamp to `t` of the inner pod.
+func (p *PodWrapper) CreationTime(t metav1.Time) *PodWrapper {
```
CreationTimestamp?

yibozhuang replied:
`CreationTimestamp ` is a Pod attribute so cannot reuse that as a function or it will conflict.

kerthcet replied:
`p.ObjectMeta.CreationTimestamp = t`  works.

yibozhuang replied:
Yes you are right, sorry, fixed now.

kerthcet commented:
thanks @yibozhuang for your wonderful work, and I left some comments. Maybe @Huang-Wei will take a further review after.

sanposhiho commented on pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go:
```diff
@@ -54,7 +55,6 @@ func createPodWithAffinityTerms(namespace, nodeName string, labels map[string]st
 			},
```
You can also use pod wrapper in `createPodWithAffinityTerms`.

yibozhuang replied:
Thanks, converted `createPodWithAffinityTerms` to use `PodWrapper`

sanposhiho commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -58,30 +58,9 @@ var (
+	highPriorityPodInfo     = framework.NewPodInfo(st.MakePod().Name("hpp").Namespace("ns1").UID("hppns1").Priority(highPriority).Obj())
+	highPriNominatedPodInfo = framework.NewPodInfo(st.MakePod().Name("hpp").Namespace("ns1").UID("hppns1").Priority(highPriority).NominatedNodeName("node1").Obj())
+	medPriorityPodInfo      = framework.NewPodInfo(&v1.Pod{
 		ObjectMeta: metav1.ObjectMeta{
```
Please use PodWrapper as well for medPriorityPodInfo

yibozhuang replied:
`medPriorityPodInfo` and `unschedulablePodInfo` have annotations which requires #109536 to be merged first since that adds support for Pod annotations so I think that can be a separate PR or I can rebase after.

sanposhiho replied:
But you've already implemented `CreationTimestamp` in this PR which is included in https://github.com/kubernetes/kubernetes/pull/109536, right?
https://github.com/kubernetes/kubernetes/pull/109536/files#diff-b8c43b6841e30bdec0f6bc6436e7d4c1bb3a67073dfc7efafcc4522429458961R214
So, I think you need to rebase after #109536 merged regardless of this change.

yibozhuang replied:
Yea that makes sense. I've updated to use PodWrapper for `medPriorityPodInfo` and `unschedulablePodInfo` and will rebase.

sanposhiho commented on pkg/scheduler/testing/wrappers.go:
```diff
@@ -417,6 +446,20 @@ func (p *PodWrapper) Label(k, v string) *PodWrapper {
 	return p
 }
 
+// MetadataLabels sets all {k,v} pair provided by `labels` to the inner pod.
+func (p *PodWrapper) MetadataLabels(labels map[string]string) *PodWrapper {
```
Given `Label` method is exist, we should name this as `Labels`?

yibozhuang replied:
`Labels` is a Pod attribute so cannot reuse that as a function or it will conflict.

sanposhiho replied:
Really? I think we can because we already have `Name` method on PodWrapper.

Huang-Wei replied:
> Labels is a Pod attribute so cannot reuse that as a function or it will conflict.

We can actually :) In that case we should specify the full path rather than emitting the anonymous field (ObjectMeta):

```go
// Label sets a {k,v} pair to the inner pod.
func (p *PodWrapper) Label(k, v string) *PodWrapper {
	if p.ObjectMeta.Labels == nil {
		p.ObjectMeta.Labels = make(map[string]string)
	}
	p.ObjectMeta.Labels[k] = v
	return p
}

// Labels sets all {k,v} pair provided by `labels` to the inner pod.
func (p *PodWrapper) Labels(labels map[string]string) *PodWrapper {
	for k, v := range labels {
		p.Label(k, v)
	}
	return p
}
```

yibozhuang replied:
Got it, thanks!
Fixed.

yibozhuang commented on pkg/scheduler/extender_test.go:
```diff
@@ -352,56 +351,23 @@ func TestIsInterested(t *testing.T) {
-				},
-			},
+			pod: st.MakePod().Container("app").InitContainerReq(map[v1.ResourceName]string{
+				"memory": "100Mi",
```
fixed.

yibozhuang commented on pkg/scheduler/extender_test.go:
```diff
@@ -352,56 +351,23 @@ func TestIsInterested(t *testing.T) {
-				},
-			},
+			pod: st.MakePod().Req(map[v1.ResourceName]string{
+				"memory": "100Mi",
```
Thanks, changed to "0".

yibozhuang commented on pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go:
```diff
@@ -94,7 +94,7 @@ func TestRequiredAffinitySingleNode(t *testing.T) {
 					},
 				}, nil),
-			pods: []*v1.Pod{{Spec: v1.PodSpec{NodeName: "machine1"}, ObjectMeta: metav1.ObjectMeta{Labels: podLabel}}},
+			pods: []*v1.Pod{st.MakePod().Label("service", "securityscan").Node("machine1").Obj()},
```
Sure thing, done.

yibozhuang commented on pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go:
```diff
@@ -2242,32 +2219,12 @@ func TestPreFilterStateClone(t *testing.T) {
-		return &v1.Pod{
-			ObjectMeta: metav1.ObjectMeta{Name: "normal", Labels: labelMap},
-			Spec:       v1.PodSpec{NodeName: "nodeA"},
+			pod = pod.Label(l, "")
```
Thanks, done.

yibozhuang commented on pkg/scheduler/framework/plugins/noderesources/balanced_allocation_test.go:
```diff
@@ -21,6 +21,8 @@ import (
 	"reflect"
 	"testing"
 
+	st "k8s.io/kubernetes/pkg/scheduler/testing"
```
Fixed.

yibozhuang commented on pkg/scheduler/framework/plugins/noderesources/balanced_allocation_test.go:
```diff
@@ -21,6 +21,8 @@ import (
+
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Fixed.

yibozhuang commented on pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go:
```diff
@@ -362,44 +339,24 @@ func TestResourceBinPackingSingleExtended(t *testing.T) {
-				},
-			},
-		},
+	extendedResourcePod1 := map[v1.ResourceName]string{
```
The current code uses `extendedResourcePod1` since these are pod resources so I think keeping the name as it is makes more sense.

kerthcet replied:
But it's a bad naming I think. `extendedResourcePod1` indicates it's a pod not the pod resource. We used to name this   `extendedResourcePod1` because it's actually a pod.

yibozhuang replied:
Sure, updated.

yibozhuang commented on pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go:
```diff
@@ -362,44 +339,24 @@ func TestResourceBinPackingSingleExtended(t *testing.T) {
-				},
-			},
-		},
+	extendedResourcePod2 := map[v1.ResourceName]string{
```
Same, using the current name makes more sense since it is pod resources.

yibozhuang replied:
Updated.

yibozhuang commented on pkg/scheduler/framework/plugins/nodevolumelimits/non_csi_test.go:
```diff
@@ -876,179 +599,47 @@ func TestEBSLimits(t *testing.T) {
-			},
-		},
-	}
+	deletedPVPod2 := st.MakePod().PVC("pvcWithDeletedPV").Obj()
```
Sure thing, did the refactor to move all the pod using PVCs to the top and have all tests consume them.

yibozhuang commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:
```diff
@@ -249,44 +218,8 @@ func TestISCSIDiskConflicts(t *testing.T) {
 
-	podWithReadWriteOncePodPVC := &v1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			// Required for querying lister for PVCs in the same namespace.
```
Done.

yibozhuang commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:
```diff
@@ -249,44 +218,8 @@ func TestISCSIDiskConflicts(t *testing.T) {
-	}
-	podWithReadWriteManyPVC := &v1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			// Required for querying lister for PVCs in the same namespace.
```
Done.

yibozhuang commented:
/retest

Huang-Wei commented on pkg/scheduler/testing/wrappers.go:
```diff
@@ -285,6 +297,24 @@ func (p *PodWrapper) HostPort(port int32) *PodWrapper {
+}
+
+// ContainerPortAtIndex sets the ports with value `ports` for the container at `index`
+func (p *PodWrapper) ContainerPortAtIndex(ports []v1.ContainerPort, index int) *PodWrapper {
```
Not sure I like it as `index` may be out of bound.

Is there any case `ContainerPort` cannot fit?

yibozhuang replied:
The only case where `ContainerPort` doesn't fit this `MakePod` at https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/internal/cache/cache_test.go#L1739

Since when we call `Req()` under `PodWrapper` it creates the container and we want to set the port for that container rather than create and append.

But I am doing something like this for now to have it fit using `ContainerPort`: https://github.com/kubernetes/kubernetes/pull/109684/files#diff-c93487173ed1e8915cdcdd46130ed723d517e191c35e817cfd3f76c913b94281R1707-R1711

Huang-Wei replied:
I feel like a ContainerWrapper is worthy, similar to NodeSelectorWrapper as a "second-level" wrapper, which is passed PodWrapper:

```go
st.MakePod().Containers(st.MakeContainer().Req().Port().Obj(), ...)
```

yibozhuang replied:
Sure thing, created `ContainerWrapper` and refactored the existing `PodWrapper` code to leverage this. I agree that this does look cleaner.

Huang-Wei commented on pkg/scheduler/testing/wrappers.go:
```diff
@@ -438,6 +550,27 @@ func (p *PodWrapper) Req(resMap map[v1.ResourceName]string) *PodWrapper {
 }
 
+// InitContainerReq adds a new init container to the inner pod with given resource map.
+func (p *PodWrapper) InitContainerReq(resMap map[v1.ResourceName]string) *PodWrapper {
```
Prefer `InitReq`

yibozhuang replied:
Fixed.

Huang-Wei reviewed (commented):
Thanks Yibo. Some comments below.

yibozhuang commented on pkg/scheduler/testing/wrappers.go:
```diff
@@ -285,6 +297,24 @@ func (p *PodWrapper) HostPort(port int32) *PodWrapper {
+	p.Spec.Containers = append(p.Spec.Containers, v1.Container{
+		Ports: ports,
+	})
+	return p
```
Thanks, fixed.

sanposhiho reviewed (approved):
Looks good to me. 
Let's`/hold` this until https://github.com/kubernetes/kubernetes/pull/109536 is merged.

/lgtm 
/hold

Huang-Wei commented:
/approve

ahg-g reviewed (commented):
This is a great cleanup, thanks for doing it.

This is hard to review in detail, but can you confirm that no test in this PR changed semantically?

Huang-Wei commented on pkg/scheduler/extender_test.go:
```diff
@@ -424,15 +390,15 @@ func TestConvertToMetaVictims(t *testing.T) {
 					Pods: []*v1.Pod{
-						{ObjectMeta: metav1.ObjectMeta{Name: "pod1", UID: "uid1"}},
-						{ObjectMeta: metav1.ObjectMeta{Name: "pod3", UID: "uid3"}},
+						st.MakePod().Name("pod1").UID("uid1").Obj(),
```
we can make MakePod() accept variadic parameters so as to avoid significant changes.

> it seems that in many places we create pods without names

yup, UTs don't really need a pod name.

yibozhuang replied:
> we can make MakePod() accept variadic parameters so as to avoid significant changes.
What would the `MakePod()` look like for this case?

Are you imagining something like?

```go
func MakePod(name ...string) *PodWrapper {
	wrapper := &PodWrapper{v1.Pod{}}
	if len(name) >= 1 {
                wrapper.SetName(name[0])
        }
        return wrapper
}
```

This will be kind of weird if the caller is specifying multiple values as only the first will be used.

> yup, UTs don't really need a pod name.
I agree here that UTs don't necessarily need specify a name so enforcing may not be ideal.

Maybe we can consider doing a separate clean up for this case in a different PR?

Huang-Wei replied:
I'm fine with leaving it as a follow-up.

ahg-g replied:
sgtm

yibozhuang committed:
cleanup: move scheduler tests to use PodWrapper

Move to use testing PodWrapper where applicable to
reduce duplicating pod creation code and shorten
number of lines.

Adding additional wrapper functions in PodWrapper
to ensure it covers all pod spec under tests.

Signed-off-by: Yibo Zhuang <yibzhuang@gmail.com>

yibozhuang committed:
cleanup: move scheduler plugin tests to use PodWrapper

Move scheduler plugin unit tests use testing PodWrapper
where applicable to reduce duplicating pod creation
code and shorten number of lines.

Signed-off-by: Yibo Zhuang <yibzhuang@gmail.com>

yibozhuang commented on pkg/scheduler/framework/plugins/noderesources/balanced_allocation_test.go:
```diff
@@ -182,15 +167,15 @@ func TestNodeResourcesBalancedAllocation(t *testing.T) {
-				{Spec: machine1Spec, ObjectMeta: metav1.ObjectMeta{Labels: labels1}},
-				{Spec: machine2Spec, ObjectMeta: metav1.ObjectMeta{Labels: labels1}},
-				{Spec: machine2Spec, ObjectMeta: metav1.ObjectMeta{Labels: labels1}},
+				st.MakePod().Node("machine1").Labels(labels2).Obj(),
```
sure thing, renamed all `machine` naming to `node`

yibozhuang commented on pkg/scheduler/framework/plugins/helper/spread_test.go:
```diff
@@ -48,16 +49,7 @@ func TestGetPodServices(t *testing.T) {
-				},
-			},
-		}
+		pod := st.MakePod().Name(fmt.Sprintf("test-pod-%d", i)).Namespace("test").Label("app", fmt.Sprintf("test-%d", i)).Label("label", fmt.Sprintf("label-%d", i)).Obj()
```
sure thing, done.

yibozhuang commented:
> This is hard to review in detail, but can you confirm that no test in this PR changed semantically?

Yea I confirmed as I was doing the clean up that the pod spec after the change matches with the previous pod spec.

yibozhuang commented:
/unhold

yibozhuang commented:
/retest

Huang-Wei commented:
/lgtm

yibozhuang commented:
/retest
